Hybrid Gateway guide (#5676)
solo-io/gloo: docs/content/guides/traffic_management/listener_configuration/hybrid_gateway.md

---
title: Hybrid Gateway
weight: 10
description: Define multiple HTTP or TCP Gateways within a single Gateway CRD
---

> Hybrid Gateways allow you to define multiple HTTP or TCP Gateways for a single Gateway CRD with distinct matching criteria.

Reviewer: I think it would be valuable to add more background on what problem hybrid gateway's solve. Why would I benefit from using one? I think outlining that a HybridGateway gives you all the functionality of the existing HttpGateway and TcpGateway plus the ability to dynamically opt into each of them based on criteria (that we can explicitly list out) will help users understand the situations they might want to leverage one.

Reviewer: I also think we could speak more about how this HybridGateway allows users more control over the FilterChainMatching.

---

## Only accept requests from a particular CIDR range

> Hybrid Gateways allow us to treat traffic from particular IPs differently.

Reviewer: Perhaps this is unnecessary, but I think we could provide even more context around why we would want to treat different IPs differently. ie we could outline a situation where 
```
A certain service wishes to enforce security best practices for most connections. For example, only allow TLS >= 1.2, and a curated list of ciphers. Unfortunately, some legacy clients have not yet migrated, and they must still use TLS1.1. We want to allow TLS1.1 connections only from these clients. Crucially these originate from the same domain, so the most trivial client discriminator that we can have is the source IP.
```

Reviewer 2: +1 i think some more context might be handy

Author: The two lines below were meant to add that context
Is there more that can/should be added?

In this example we will demonstrate how to only allow requests from one IP to reach an upstream while short-circuiting all other IPs with a direct response action.

We will pick up where the [Hello World guide]({{< versioned_link_path fromRoot="/guides/traffic_management/hello_world" >}}) leaves off.

> To start we will add a second VirtualService that also matches all requests and has a directResponseAction:

Reviewer: Can we either update the comment 'matches all requests' to clarify that it's the `all-pets` endpoint, or just update the VS to match all paths?

Author: I updated it everywhere but the VS definition 🤦‍♂️

```bash
kubectl edit -n gloo-system gateway gateway-proxy
```

{{< highlight yaml "hl_lines=7-21" >}}
apiVersion: gateway.solo.io/v1
kind: Gateway
metadata: # collapsed for brevity
spec:
  bindAddress: '::'
  bindPort: 8080
  hybridGateway:
    matchedGateways:
      - httpGateway:
          virtualServices:
            - name: default
              namespace: gloo-system
        matcher:
          sourcePrefixRanges:
            - addressPrefix: 0.0.0.0
              prefixLen: 1
      - httpGateway:
          virtualServices:
            - name: client-ip-reject
              namespace: gloo-system
        matcher: {}
  proxyNames:
  - gateway-proxy
  useProxyProto: false
status: # collapsed for brevity
{{< /highlight >}}

Note: We use a range of 0.0.0.0/1 in order to have a high chance of matching the client's IP without knowing it specifically. A different and/or narrower range may be used if we know more about the client's IP.

This results in a proxy that looks like:

```yaml

> apiVersion: gloo.solo.io/v1

Reviewer: Given that this is an intermediate resource, do we need to expose this to the user? If so, is there specific config that is relevant to them that we can highlight or direct them to?

Author: Modeled on [the tcp proxy guide](https://docs.solo.io/gloo-edge/latest/guides/traffic_management/listener_configuration/tcp_proxy/), but I suppose isn't really relevant here
If you think it can go, I'm on board

Reviewer: Yeah I think yaml should provide guidance for an action to take, and since there is no action taken as a result of this Proxy yaml, I think we should just remove it.

---
title: Hybrid Gateway
weight: 10

> description: Define multiple HTTP or TCP Gateways within a single Gateway CRD

Reviewer: Let's either refer to this as a Gateway CR or just Gateway. The CRD is the definition of what a Gateway can be (https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/) whereas the CR (https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/) is what a specific gateway is

---

> Hybrid Gateways allow users to define multiple HTTP or TCP Gateways for a single Gateway CRD with distinct matching criteria. 

Reviewer: same as above: let's use Gateway CR or just Gateway

---

> Hybrid Gateways provide all of the functionality of HTTP and TCP Gateways with the added ability to dynamically select which Gateway a given request is routed to based on request properties.

Reviewer 2: ```suggestion
Hybrid gateways expand the functionality of HTTP and TCP gateways by letting you use request properties to choose which gateway the request routes to.
```

Reviewer 2: I think something should be added in this top section too about how this allows you to expose multiple HTTP and/or TCP gateways on a single port, seems important to highlght

Selection is done based on `Matcher` fields, which map to a subset of Envoy [`FilterChainMatch`](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/listener/v3/listener_components.proto#config-listener-v3-filterchainmatch) fields.

## Only accept requests from a particular CIDR range

Hybrid Gateways allow us to treat traffic from particular IPs differently.
One case where this might come in handy is if a set of clients are at different stages of migrating to TLS >=1.2 support, and therefore we want to enforce different TLS requirements depending on the client.
If the clients originate from the same domain, it may be necessary to dynamically route traffic to the appropriate Gateway based on source IP.

> In this example we will demonstrate how to only allow requests from one IP to reach an upstream while short-circuiting all other IPs with a direct response action.

Reviewer 2: ```suggestion
In this example, we will allow requests only from one IP to reach an upstream, while short-circuiting requests from all other IPs by using a direct response action.
```

> We will pick up where the [Hello World guide]({{< versioned_link_path fromRoot="/guides/traffic_management/hello_world" >}}) leaves off.

Reviewer 2: just for some visual variety to signal where the actual task starts
```suggestion
**Before you begin**: Complete the [Hello World guide]({{< versioned_link_path fromRoot="/guides/traffic_management/hello_world" >}}) demo setup.
```

To start we will add a second VirtualService that also matches all requests and has a directResponseAction:

```yaml
kubectl apply -n gloo-system -f - <<EOF

> apiVersion: gateway.solo.io/v1

Reviewer: This VS is only used in the second example, should we move the resource definition closer to the steps that rely on it?

Author: Even though we don't request it til later, we refer to it in the next step when we edit/create the Hybrid Gateway
While we could make a Hybrid Gateway with just the one matched gateway consisting of the petstore VS, I think it's meaningful to show that the routing is making a deterministic selection between multiple Gateways

kind: VirtualService
metadata:
  name: 'client-ip-reject'
  namespace: 'gloo-system'
spec:
  virtualHost:
    domains:
      - '*'
    routes:
      - matchers:
          - prefix: /
        directResponseAction:
          status: 403
          body: "client ip forbidden\n"
EOF
```

> Next let's update the existing `gateway-proxy` Gateway CRD, replacing the default `httpGateway` with a [`hybridGateway`]({{< versioned_link_path fromRoot="/reference/api/github.com/solo-io/gloo/projects/gateway/api/v1/gateway.proto.sk/#hybridgateway" >}}) as follows:

Reviewer: crd -> cr

```bash
kubectl edit -n gloo-system gateway gateway-proxy
```

{{< highlight yaml "hl_lines=7-21" >}}
apiVersion: gateway.solo.io/v1
kind: Gateway
metadata: # collapsed for brevity
spec:
  bindAddress: '::'
  bindPort: 8080
  hybridGateway:
    matchedGateways:
      - httpGateway:
          virtualServices:
            - name: default
              namespace: gloo-system
        matcher:
          sourcePrefixRanges:
            - addressPrefix: 0.0.0.0
              prefixLen: 1
      - httpGateway:
          virtualServices:
            - name: client-ip-reject
              namespace: gloo-system
        matcher: {}
  proxyNames:
  - gateway-proxy
  useProxyProto: false
status: # collapsed for brevity
{{< /highlight >}}

> Note: We use a range of 0.0.0.0/1 in order to have a high chance of matching the client's IP without knowing it specifically. A different and/or narrower range may be used if we know more about the client's IP.

Reviewer 2: ```suggestion
Note: The range of 0.0.0.0/1 provides a high chance of matching the client's IP without knowing the specific IP. If you know more about the client's IP, you can specify a different, narrower range.
```

                    metadata: # collapsed for brevity
                    options:
                      prefixRewrite: /api/pets
                    routeAction:
                      single:
                        upstream:
                          name: default-petstore-8080
                          namespace: gloo-system
          matcher:
            sourcePrefixRanges:
              - addressPrefix: 0.0.0.0
                prefixLen: 1
        - httpListener:
            virtualHosts:
              - domains:
                  - '*'
                metadata: # collapsed for brevity
                name: gloo-system.client-ip-reject
                routes:
                - directResponseAction:
                    body: |
                      client ip forbidden
                    status: 403
                  matchers:
                  - prefix: /
                  metadata: # collapsed for brevity
            matcher: {}
    metadata: # collapsed for brevity
    name: listener-::-8080
    useProxyProto: false
  - bindAddress: '::'
    bindPort: 8443
    httpListener: {}
    metadata: # collapsed for brevity
status: # collapsed for brevity
```

> We can make a request to the proxy and will find that we get the `200` response:

Reviewer 2: ```suggestion
Make a request to the proxy, which returns a `200` response because the client IP address matches to the 0.0.0.0/1 range:
```

```bash
$ curl "$(glooctl proxy url)/all-pets"
[{"id":1,"name":"Dog","status":"available"},{"id":2,"name":"Cat","status":"pending"}]
```

> Also observe that if we make a request to an endpoint not matched by the `default` VirtualService we get a `404` response and _do not_ hit the `client-ip-reject` VirtualService:

Reviewer 2: ```suggestion
Note that a request to an endpoint that is not matched by the `default` VirtualService returns a `404` response, and the request _does not_ hit the `client-ip-reject` VirtualService:
```

```bash
$ curl -i "$(glooctl proxy url)/foo"
HTTP/1.1 404 Not Found
date: Tue, 07 Dec 2021 17:48:49 GMT
server: envoy
content-length: 0
```
This is because the `Matcher`s in the `HybridGateway` determine which `MatchedGateway` a request will be routed to, regardless of what routes that gateway has.

> ### Observe that request from unmatched IP hits catchall gateway 

Reviewer 2: ```suggestion
### Route requests from non-matching IPs to a catchall gateway 
```

> If we update the matcher to have a specific IP range that our client's IP is not a member of, we will expect our request to miss the matcher and fall through to the catchall gateway which is configured to respond `403`.

Reviewer 2: ```suggestion
Next, update the matcher to use a specific IP range that our client's IP is not a member of. Requests from this client IP will now skip this matcher, and will instead match to a catchall gateway that is configured to respond with `403`.
```

```bash
kubectl edit -n gloo-system gateway gateway-proxy
```
{{< highlight yaml "hl_lines=15-16" >}}
apiVersion: gateway.solo.io/v1
kind: Gateway
metadata: # collapsed for brevity
spec:
  bindAddress: '::'
  bindPort: 8080
  hybridGateway:
    matchedGateways:
      - httpGateway:
          virtualServices:
            - name: default
              namespace: gloo-system
        matcher:
          sourcePrefixRanges:
            - addressPrefix: 1.2.3.4
              prefixLen: 32
      - httpGateway:
          virtualServices:
            - name: client-ip-reject
              namespace: gloo-system
        matcher: {}
  proxyNames:
  - gateway-proxy
  useProxyProto: false
status: # collapsed for brevity
{{< /highlight >}}

The Proxy will update accordingly.

> We can now make a request to the proxy and will find that we get the `403` response for any endpoint:

Reviewer 2: ```suggestion
Make a request to the proxy, which now returns a `403` response for any endpoint:
```
